Store digests into gocli binary

qinqon opened the pull request:
As part of start moving from bash to gocli and also a future release pipeline
that will release the gocli binary with digest on it, this commit read
image digest from cluster-provision/gocli/images.json and store it
at the gocli binary.

This way a cluster can simply be started with `./cli run k8s-1.18` for example also
it allow us to remove images.sh from bash scripts.

To also support development the flag `--container-suffix` allow
to override the digest stored at the binary making possible to check local provisioned
doing `./cli run --container-registry="" --container-prefix=:latest k8s-1.18`.

Signed-off-by: Quique Llorente <ellorent@redhat.com>

qinqon commented:
/cc @danielBelenky

qinqon commented:
/test check-provision-k8s-1.18

qinqon commented:
/test push-gocli

qinqon commented:
/test check-provision-k8s-1.18

qinqon commented:
/test check-provision-k8s-1.16 check-provision-k8s-1.17 check-provision-k8s-1.18

qinqon commented:
/test push-gocli

qinqon commented:
/test check-provision-k8s-1.16 check-provision-k8s-1.17 check-provision-k8s-1.18

qinqon commented:
/test check-provision-k8s-1.18 
/test check-provision-k8s-1.17
/test check-provision-k8s-1.16

qinqon commented:
/test push-gocli

qinqon commented:
/test check-provision-k8s-1.18
/test check-provision-k8s-1.17
/test check-provision-k8s-1.16

qinqon commented:
/test push-gocli

danielBelenky commented on cluster-provision/gocli/Makefile:
```diff
@@ -1,20 +1,20 @@
-	CGO_ENABLED=0 GOOS=linux GOARCH=amd64 $(GO) build -o $(BIN_DIR)/cli ./cmd/cli
-
+cli:
+	CGO_ENABLED=0 GOOS=linux GOARCH=amd64 $(GO) build -ldflags "-X 'kubevirt.io/kubevirtci/cluster-provision/gocli/images.SUFFIXES=$(shell cat $(IMAGES_FILE)|base64 -w 0)'" -o $(BIN_DIR)/cli ./cmd/cli
```
Why you prefer to use `ldflags` instead of hard-coding this in as constants or something in a dedicated constants module?

qinqon replied:
The idea is to not store hashes at repo at all when release pipeline is implemented, the release pipeline will just bump whatever digest is need to be bump and pass a json file to this.

danielBelenky commented:
Looks good, but there's a merge conflict. Can you resolve it so we can merge this?

qinqon commented:
/test push-gocli

qinqon commented:
@danielBelenky added following changes
- rebased
- updated gocli digest (now we have to do so everytime we create update providers)
- updated contributing guide
- updated `make bump`

qinqon commented:
/test check-provision-k8s-1.18

qinqon commented:
/retest

qinqon commented:
/test check-provision-k8s-1.19
/test check-provision-k8s-1.18 
/test check-provision-k8s-1.17
/test check-provision-k8s-1.16

qinqon commented:
Provisioning is failing in unrelated issue 
```
-- Unit docker.service has begun starting up.
Sep 28 11:42:31 node01 dockerd[31572]: time="2020-09-28T11:42:31.563428623Z" level=info msg="Starting up"
Sep 28 11:42:31 node01 dockerd[31572]: time="2020-09-28T11:42:31.564862329Z" level=info msg="parsed scheme: \"unix\"" module=grpc
Sep 28 11:42:31 node01 dockerd[31572]: time="2020-09-28T11:42:31.564889499Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
Sep 28 11:42:31 node01 dockerd[31572]: time="2020-09-28T11:42:31.564909311Z" level=info msg="ccResolverWrapper: sending update to cc: {[{unix:///run/containerd/containerd.sock 0  <nil>}] <nil>}" module=grpc
Sep 28 11:42:31 node01 dockerd[31572]: time="2020-09-28T11:42:31.564918766Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
Sep 28 11:42:31 node01 dockerd[31572]: time="2020-09-28T11:42:31.569710015Z" level=info msg="parsed scheme: \"unix\"" module=grpc
Sep 28 11:42:31 node01 dockerd[31572]: time="2020-09-28T11:42:31.569741594Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
Sep 28 11:42:31 node01 dockerd[31572]: time="2020-09-28T11:42:31.569760680Z" level=info msg="ccResolverWrapper: sending update to cc: {[{unix:///run/containerd/containerd.sock 0  <nil>}] <nil>}" module=grpc
Sep 28 11:42:31 node01 dockerd[31572]: time="2020-09-28T11:42:31.569770220Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
Sep 28 11:42:31 node01 dockerd[31572]: time="2020-09-28T11:42:31.571539550Z" level=error msg="Failed to built-in GetDriver graph btrfs /var/lib/docker"
Sep 28 11:42:31 node01 systemd[1]: Started man-db-cache-update.service.
-- Subject: Unit man-db-cache-update.service has finished start-up
-- Defined-By: systemd
-- Support: https://access.redhat.com/support
-- 
-- Unit man-db-cache-update.service has finished starting up.
-- 
-- The start-up result is RESULT.
Sep 28 11:42:31 node01 dockerd[31572]: time="2020-09-28T11:42:31.799638729Z" level=warning msg="Your kernel does not support cgroup blkio weight"
Sep 28 11:42:31 node01 dockerd[31572]: time="2020-09-28T11:42:31.799663249Z" level=warning msg="Your kernel does not support cgroup blkio weight_device"
Sep 28 11:42:31 node01 dockerd[31572]: time="2020-09-28T11:42:31.799808165Z" level=info msg="Loading containers: start."
Sep 28 11:42:31 node01 kernel: bridge: filtering via arp/ip/ip6tables is no longer available by default. Update your scripts to load br_netfilter if you need this.
Sep 28 11:42:31 node01 kernel: Bridge firewalling registered
Sep 28 11:42:31 node01 dockerd[31572]: time="2020-09-28T11:42:31.820624574Z" level=warning msg="Failed to find iptables: exec: \"iptables\": executable file not found in $PATH"
Sep 28 11:42:31 node01 dockerd[31572]: failed to start daemon: Error initializing network controller: error obtaining controller instance: failed to create NAT chain DOCKER: Iptables not found
Sep 28 11:42:31 node01 systemd[1]: docker.service: Main process exited, code=exited, status=1/FAILURE
Sep 28 11:42:31 node01 systemd[1]: docker.service: Failed with result 'exit-code'.
Sep 28 11:42:31 node01 systemd[1]: Failed to start Docker Application Container Engine.
-- Subject: Unit docker.service has failed
```

qinqon commented:
/test check-provision-k8s-1.19
/test check-provision-k8s-1.18
/test check-provision-k8s-1.17
/test check-provision-k8s-1.16

dhiller reviewed (approved):
/approve
/hold until the shas have been updated

qinqon commented:
/test push-k8s-1.17
/test push-k8s-1.18
/test push-k8s-1.19

dhiller reviewed (approved):
/hold cancel
